### PR TITLE
Permissions Boundary Infrastructure

### PR DESCRIPTION
**Pull request by claytondaley**
This is an implementation of #464 (where we're still discussing the proposal).  I've never tried to run tests cases, but the PR does lint.

### CODE REVIEW

**Comment by andreaswittig** [comment]
I see, that making use of permission boundaries helps to avoid privilege escalation from whoever needs to deploy the CloudFormation templates. However, attaching a permission boundary policy to one of the IAM roles might lead to a broken deployment. Because the permission boundary might restrict necessary actions.

I'm not sure, wether we should add that feature. It seems to me, that only very few organisations make use of permission boundaries. Not sure if it is worth the extra complexity.

**Comment by claytondaley** [comment]
I'm not proposing a default boundary so 99% of users will see no difference.  To the extent that an organization has a boundary, it prevents that user from deploying the particular resources *in any way* -- not just CF and not just these templates.

For example a user who cannot deploy a NAT Gateway simply cannot do so.  Your base `vpc` template will never work for them, but failing to support a `PermissionsBoundary` doesn't help that user.  That user *can* create a VPC template that exposes the same interface but does not use a NAT Gateway and they could continue to use the rest of your modules (so long as the relevant ones support a `PermissionsBoundary` and don't require other restricted resources).  For unrelated reasons we actually use a custom VPC template that is all-private (no internet access) but exposes the same interface so we already have a proof-of-concept of this approach.

File: security/cloudtrail.yaml [hunk 1]
```diff
@@ -17,13 +17,21 @@ Description: 'Security: CloudTrail across all regions, a cloudonaut.io template'
 Metadata:
   'AWS::CloudFormation::Interface':
     ParameterGroups:
+    - Label:
```
Comment (michaelwittig) [inline security/cloudtrail.yaml:20]: can we move this to the end?

File: ecs/cluster-cost-optimized.yaml [hunk 2]
```diff
@@ -28,6 +28,10 @@ Metadata:
       - ParentClientStack1
       - ParentClientStack2
       - ParentClientStack3
+    - Label:
```
Comment (michaelwittig) [inline ecs/cluster-cost-optimized.yaml:31]: can we move this to the end?

File: ecs/cluster.yaml [hunk 3]
```diff
@@ -28,6 +28,10 @@ Metadata:
       - ParentClientStack1
       - ParentClientStack2
       - ParentClientStack3
+    - Label:
```
Comment (michaelwittig) [inline ecs/cluster.yaml:31]: can we move this to the end?

File: ecs/service-cluster-alb.yaml [hunk 4]
```diff
@@ -23,6 +23,10 @@ Metadata:
       - ParentClusterStack
       - ParentAlertStack
       - ParentZoneStack
+    - Label:
```
Comment (michaelwittig) [inline ecs/service-cluster-alb.yaml:26]: can we move this to the end?

File: ecs/service-dedicated-alb.yaml [hunk 5]
```diff
@@ -26,6 +26,10 @@ Metadata:
       - ParentAlertStack
       - ParentZoneStack
       - ParentS3StackAccessLog
+    - Label:
```
Comment (michaelwittig) [inline ecs/service-dedicated-alb.yaml:29]: can we move this to the end?

File: fargate/service-cloudmap.yaml [hunk 6]
```diff
@@ -29,6 +29,10 @@ Metadata:
       - ParentClientStack2
       - ParentClientStack3
       - ParentSSHBastionStack
+    - Label:
```
Comment (michaelwittig) [inline fargate/service-cloudmap.yaml:32]: can we move this to the end?

File: fargate/service-cluster-alb.yaml [hunk 7]
```diff
@@ -27,6 +27,10 @@ Metadata:
       - ParentClientStack1
       - ParentClientStack2
       - ParentClientStack3
+    - Label:
```
Comment (michaelwittig) [inline fargate/service-cluster-alb.yaml:30]: can we move this to the end?

File: fargate/service-dedicated-alb.yaml [hunk 8]
```diff
@@ -29,6 +29,10 @@ Metadata:
       - ParentClientStack1
       - ParentClientStack2
       - ParentClientStack3
+    - Label:
```
Comment (michaelwittig) [inline fargate/service-dedicated-alb.yaml:32]: can we move this to the end?

File: jenkins/jenkins2-ha-agents.yaml [hunk 9]
```diff
@@ -26,6 +26,10 @@ Metadata:
       - ParentAlertStack
       - ParentS3StackAccessLog
       - ParentZoneStack
+    - Label:
```
Comment (michaelwittig) [inline jenkins/jenkins2-ha-agents.yaml:29]: can we move this to the end?

File: jenkins/jenkins2-ha.yaml [hunk 10]
```diff
@@ -26,6 +26,10 @@ Metadata:
       - ParentAlertStack
       - ParentS3StackAccessLog
       - ParentZoneStack
+    - Label:
```
Comment (michaelwittig) [inline jenkins/jenkins2-ha.yaml:29]: can we move this to the end?

File: operations/alb-access-logs-anonymizer.yaml [hunk 11]
```diff
@@ -22,6 +22,10 @@ Metadata:
       Parameters:
       - ParentS3Stack
       - ParentAlertStack
+    - Label:
```
Comment (michaelwittig) [inline operations/alb-access-logs-anonymizer.yaml:25]: can we move this to the end?

File: operations/cloudfront-access-logs-anonymizer.yaml [hunk 12]
```diff
@@ -22,6 +22,10 @@ Metadata:
       Parameters:
       - ParentS3Stack
       - ParentAlertStack
+    - Label:
```
Comment (michaelwittig) [inline operations/cloudfront-access-logs-anonymizer.yaml:25]: can we move this to the end?

File: security/account-password-policy.yaml [hunk 13]
```diff
@@ -17,6 +17,10 @@ Description: 'Security: Account Password Policy, a cloudonaut.io template'
 Metadata:
   'AWS::CloudFormation::Interface':
     ParameterGroups:
+    - Label:
```
Comment (michaelwittig) [inline security/account-password-policy.yaml:20]: can we move this to the end?

File: security/auth-proxy-ha-github-orga.yaml [hunk 14]
```diff
@@ -25,6 +25,10 @@ Metadata:
       - ParentAlertStack
       - ParentS3StackAccessLog
       - ParentZoneStack
+    - Label:
```
Comment (michaelwittig) [inline security/auth-proxy-ha-github-orga.yaml:28]: can we move this to the end?

File: security/config.yaml [hunk 15]
```diff
@@ -17,12 +17,20 @@ Description: 'Security: Config setup, a cloudonaut.io template'
 Metadata:
   'AWS::CloudFormation::Interface':
     ParameterGroups:
+    - Label:
```
Comment (michaelwittig) [inline security/config.yaml:20]: can we move this to the end?

File: state/dynamodb.yaml [hunk 16]
```diff
@@ -22,6 +22,10 @@ Metadata:
       Parameters:
       - ParentAlertStack
       - ParentKmsKeyStack
+    - Label:
```
Comment (michaelwittig) [inline state/dynamodb.yaml:25]: can we move this to the end?

File: state/elasticache-memcached.yaml [hunk 17]
```diff
@@ -25,6 +25,10 @@ Metadata:
       - ParentZoneStack
       - ParentSSHBastionStack
       - ParentAlertStack
+    - Label:
```
Comment (michaelwittig) [inline state/elasticache-memcached.yaml:28]: can we move this to the end?

File: static-website/lambdaedge-index-document.yaml [hunk 18]
```diff
@@ -17,6 +17,10 @@ Description: 'Static Website: Index document support in subdirectories, a cloudo
 Metadata:
   'AWS::CloudFormation::Interface':
     ParameterGroups:
+    - Label:
```
Comment (michaelwittig) [inline static-website/lambdaedge-index-document.yaml:20]: can we move this to the end?

File: vpc/vpc-flow-logs.yaml [hunk 19]
```diff
@@ -21,15 +21,25 @@ Metadata:
         default: 'Parent Stacks'
       Parameters:
       - ParentVPCStack
+    - Label:
```
Comment (michaelwittig) [inline vpc/vpc-flow-logs.yaml:24]: can we move this to the end?

File: vpc/vpc-nat-instance.yaml [hunk 20]
```diff
@@ -24,6 +24,10 @@ Metadata:
       - ParentSSHBastionStack
       - ParentAlertStack
       - ParentZoneStack
+    - Label:
```
Comment (michaelwittig) [inline vpc/vpc-nat-instance.yaml:27]: can we move this to the end?

File: vpc/vpc-ssh-bastion.yaml [hunk 21]
```diff
@@ -23,6 +23,10 @@ Metadata:
       - ParentVPCStack
       - ParentAlertStack
       - ParentZoneStack
+    - Label:
```
Comment (michaelwittig) [inline vpc/vpc-ssh-bastion.yaml:26]: can we move this to the end?

File: vpc/vpc-vpn-bastion.yaml [hunk 22]
```diff
@@ -23,6 +23,10 @@ Metadata:
       - ParentVPCStack
       - ParentAlertStack
       - ParentZoneStack
+    - Label:
```
Comment (michaelwittig) [inline vpc/vpc-vpn-bastion.yaml:26]: can we move this to the end?

File: wordpress/wordpress-ha-aurora.yaml [hunk 23]
```diff
@@ -26,6 +26,10 @@ Metadata:
       - ParentAlertStack
       - ParentS3StackAccessLog
       - ParentZoneStack
+    - Label:
```
Comment (michaelwittig) [inline wordpress/wordpress-ha-aurora.yaml:29]: can we move this to the end?

File: wordpress/wordpress-ha.yaml [hunk 24]
```diff
@@ -26,6 +26,10 @@ Metadata:
       - ParentAlertStack
       - ParentS3StackAccessLog
       - ParentZoneStack
+    - Label:
```
Comment (michaelwittig) [inline wordpress/wordpress-ha.yaml:29]: can we move this to the end?

**Comment by andreaswittig** [comment]
> I'm not proposing a default boundary so 99% of users will see no difference. To the extent that an organization has a boundary, it prevents that user from deploying the particular resources _in any way_ -- not just CF and not just these templates.

I agree! Thanks for your contribution.

**Comment by michaelwittig** [comment]
I merged this into a separate branch to run our test suite (only works with our repo, not forks), see #473

**Pull request merged**